### DISCUSSION TITLE
Issue with user JS function

### DISCUSSION THREAD

**Discussion by 18kimn** [SilentVoid13/Templater#757] [Help]
Hi,

I have a problem in that Templater is triggered on some file creations, but at other times it's not and the template as shown below is left in the file without being updated. I often have to delete and recreate the file or run `Alt+R` to trigger the template. I have a template for a daily note that looks a bit like the following:

```md
# <% tp.date.now("YYYY-MM-DD") %>

[[<% tp.date.now("YYYY-MM-DD", -1) %>| <- ]] | [[<% tp.date.now("YYYY-MM-DD", 1) %>| ->]]

<!-- the rollover-daily-todos plugin is inserted here -->

<% await tp.user.displayQuote() %>

```

The `displayQuote` function fetches a quote from a JSON file I have on disk and displays it. It does this through node's `fs.promises`, so it's async.  I see the expected result in the devtools console, just not in the file itself. I'm wondering if there is a possible race condition or something similar, where file creation and template replacement finish too far apart for the template to run. Would you be able to shed light on how templater works and the most probable routes where I could have gone wrong?

Things I've checked so far:
- user scripts are enabled
- the JS file is found as a user script
- the script logs its output to the console even when the template is not rendered
- "trigger templater on new file creation" option is enabled
- the script seems to work unfailingly outside of obsidian as well

**Comment by djsnipa1** [comment]
I am having a similar issue as well if someone could please possibly point us in the right direction. Thank you!

**Comment by frizop** [comment]
Having the same issue, I tried following the guide online docs and copied pasted the exact same results, my template _never_ renders the `msg` and always inputs `undefined` into the MD file. I've verified I can see the result in the developer tools. As with the OP, I also verified that the script is found, executed. 

Edit: I've got a few screenshots that I wonder if they tell a story that's helpful
![image](https://user-images.githubusercontent.com/3605906/148287180-ca410d1b-b347-4167-af68-4b0463be45e3.png)
![image](https://user-images.githubusercontent.com/3605906/148287245-c5cac06a-afd2-49fc-ae9b-d256488c57b6.png)

Edit: Fixed by adding a return statement.

**Comment by shabegom** [comment]
@18kimn Can you share the JS function? It would help troubleshoot the issue.

**Comment by 18kimn** [comment]
Sure, it's:

```
const displayQuote = async () => {
  const quote = await fetch('http://localhost:4100').then((res) => res.json())
  const {body, link, attribution} = quote
  console.log(quote)
  return `
  > ${body}
  >
  > [${attribution}](${link})
  `
}

module.exports = displayQuote
```

The `localhost` query returning things of the form: 

```
{
  "body": "Revolutions are the locomotives of history.",
  "link":"http://marxists.org/archive/marx/works/1850/class-struggles-france/ch03.htm#locomotives",
  "attribution":"Class Struggle in France (1850)"
}
```

This would be transformed into quotes of the form: 

```
> Revolutions are the locomotives of history.
> 
> [Class Struggle in France (1850)](http://marxists.org/archive/marx/works/1850/class-struggles-france/ch03.htm#locomotives)
```

I know this doesn't really bring us to a reproducible example; I'm unable to provide one since the issue occurs sporadically and not on every file creation. I'm fairly certain that it isn't an issue with my code, as the `console.log(quote)` always shows the quote in the Obsidian devtools console even when it doesn't appear in the note that templater generates. Any general tips on how templater works or areas where this might occur would be super helpful!

**Comment by shabegom** [comment]
If it is showing the quote in the console then there isn't an issue with retrieving it from the local server...

hmmm

just for fun, let's try using a JS command instead:

```
<%*
const quote = await tp.user.displayQuote()
tR += quote
%>
```

Also strange that it only happens sporadically. Could there be another plugin conflicting? This is a weird one for sure. It should work...
